### PR TITLE
Move OpenMP calculations into its own class

### PR DESCRIPTION
**Pull request by cjolivier01**
## Description ##

Splitting out some things from the 'tuner' branch (OMP tuning) in order to make the tuner PR more palatable when it posts (smaller, simpler).

Since OMP behavior has little to do with the engine classes themselves, and the OMP logic will get more complex, move that code into its own class.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] For user-facing API changes, API doc string has been updated.
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Intersting edge cases to note here

### CODE REVIEW

File: src/engine/thread_pool.h [hunk 1]
```diff
@@ -57,7 +57,7 @@ class ThreadPool {
 
     /*! \brief Signal event upon destruction, even for exceptions (RAII) */
     struct SetReadyOnDestroy {
-      explicit inline SetReadyOnDestroy(std::shared_ptr<SimpleEvent> event)
+      explicit inline SetReadyOnDestroy(std::shared_ptr<SimpleEvent>& event)
```
Comment (cjolivier01) [inline src/engine/thread_pool.h]: Bypass a superfluous locked increment/decrement when passing the shared pointer by value

File: src/engine/threaded_engine_perdevice.cc [hunk 2]
```diff
@@ -114,7 +114,7 @@ class ThreadedEnginePerDevice : public ThreadedEngine {
           auto ptr =
           gpu_copy_workers_.Get(ctx.dev_id, [this, ctx, is_copy, nthread]() {
             // Signify to kernel that GPU is being used,  no Kernel Launch OMP (temporary behavior)
-            Engine::Get()->set_num_omp_threads_per_worker(1);
+            OpenMP::Get()->set_enabled(false);
```
Comment (cjolivier01) [inline src/engine/threaded_engine_perdevice.cc]: For now, just "switch it off", which will just return '1' for number of OMP threads whenever OpenMP class is asked.

File: src/nnvm/legacy_op_util.cc [hunk 3]
```diff
@@ -30,8 +30,10 @@ class ParsedOpProp {
   std::vector<std::string> outputs;
   // initializer
   void Init(const NodeAttrs& attrs) {
-    std::vector<std::pair<std::string, std::string> > kwargs(
-        attrs.dict.begin(), attrs.dict.end());
+    // For performance, do a reserve first and then copy attrs.dict
+    std::vector<std::pair<std::string, std::string> > kwargs;
```
Comment (cjolivier01) [inline src/nnvm/legacy_op_util.cc]: simple, independent optimization here

File: include/mxnet/engine.h [hunk 4]
```diff
@@ -272,11 +272,6 @@ class MXNET_API Engine {
    * \return Number of OMP threads that should be used per worker
    */
   virtual int num_omp_threads_per_worker() const = 0;
-
-  /*! \brief Set the number of OMP threads that should be used per worker
-   * \param num_threads_per_worker Number of OMP threads to be used per worker
-   */
-  virtual void set_num_omp_threads_per_worker(int num_omp_threads_per_worker) = 0;
```
Comment (cjolivier01) [inline include/mxnet/engine.h]: We're going to start deciding on the number of threads in OpenMP class -- not listening to random outside sources.  Not currently used, except to disable.

File: src/engine/openmp.h [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#ifndef MXNET_ENGINE_OMP_H_
+#define MXNET_ENGINE_OMP_H_
+
+namespace mxnet {
+namespace engine {
+
+#ifdef _OPENMP
```
Comment (piiswrong) [inline src/engine/openmp.h]: if not using OPENMP the OpenMP class should still exist.
Otherwise you have to wrap every reference to it in macro

Comment (cjolivier01) [reply src/engine/openmp.h]: Ok

**Comment by cjolivier01** [comment]
Does anyone have any further issues with this PR? It's going to clear CI soon...

**Comment by cjolivier01** [comment]
Rebase is severely problematic regarding conflicts. I’ve tried many times
I the past to utilize it but it just doesn’t scale.

**Comment by szha** [comment]
If there are conflicts when rebasing on a commit, the same conflict needs to be resolved when merging.

Merged commits would also show up in the diff. Effectively, the burden of separating your changes from merged changes is shifted to the reviewers by the practice of clicking merges. It's easier for you and harder for others.

It would be idea if there's a middle-ground that doesn't incur difficulty of reviewing. Unfortunately I'm not aware of one. Feel free to share if you know better.

**Comment by szha** [comment]
And still merging 😞

**Comment by szha** [comment]
@cjolivier01 do you really consider it appropriate and suitable as committer the behavior of deleting other people's comment before reaching consensus?

Personally I think this is offensive. I'd suggest that you raise discussion with @apache/mxnet-committers and make sure to get other people's consensus before deleting another comment. Thanks.

**Comment by cjolivier01** [comment]
Sheng,

If someone chooses to make snarky, arguably unprofessional comments on my
CR that have nothing to do with the actual code, then I reserve the right
to delete it.

Thanks,

-Chris

**Comment by srochel** [comment]
The specific issue has been resolved. We do expect to build a community where everybody communicates professionally and it is safe to express opinions. Deleting comments in general is not allowed. We are working on revising [https://cwiki.apache.org/confluence/display/MXNET/Development+Process](url) to clarify expected  behavior and operating rules. This will include procedure for deleting illegal or inappropriate comments as an exceptional case.

**Comment by hyandell** [comment]
I've added a note on Deleting GitHub Comments to the Development Process wiki. I've mailed that to dev@ in case there's desired discussion/editing.

srochel@ - as you're not an Apache committer, I'm guessing that was the royal 'We' there. May I remind you not to speak for an Apache project simply because you employ a committer.

**Pull request merged**